src/javashopper.py

```diff
@@ -0,0 +1,111 @@
+import logging
+import os
+import time
+
+import keyboard
+import mouse
+import screen
+
+from config import Config
+from logger import Logger
+from npc_manager import NpcManager, Npc
+from template_finder import TemplateFinder, load_template
+
+
+def exit():
+    Logger.info("Exiting...")
+    os._exit(0)
+
+
+class JavaShopper:
+    """
+    Shop javazon 20 ias 3 java skill gloves
+    """
+
+    def __init__(self, config):
+        self._screen = screen.Screen()
+        self.config = config
+        self._template_finder = TemplateFinder(self._screen)
+        self._npc_manager = NpcManager(
+            screen=self._screen, template_finder=self._template_finder
+        )
+        keyboard.add_hotkey(self.config.general["exit_key"], exit)
+
+    def run(self):
+        Logger.info("STARTING JAVAZON GG GLOVES SHOPPER!")
+
+        # Are we at the right start place?
+        t1 = time.time()
+        while True:
+            success, pos = self._template_finder.search_and_wait("A5_RED_PORTAL")
+            if success:
+                mouse.move(*pos, duration=0.1)
+                mouse.click()
+                break
+
+            if time.time() > t1 + 10:
+                Logger.info("Cannot find start place :( exiting...")
+                os._exit(0)
+            time.sleep(3)
+
+        self.shop_loop()
+
+    def shop_loop(self):
+        while True:
+            time.sleep(3.6)
+            success, pos = self._template_finder.search_and_wait("A5_RED_PORTAL")
+            mouse.move(*pos, duration=0.1)
+            mouse.click()
+            time.sleep(1)
```

Comment:
This  sometimes failed for me either because loading time was too long and there is still a loading screen, or char clicks too quick on tp and say something that he can not do this. Also if you don't clear the skeletons they can cause issues. Here is what worked out quite robust for me:
```python
def wait_for_loading_screen(screen: Screen, time_out):
    start = time.time()
    while time.time() - start < time_out:
        img = screen.grab()
        # TODO: Take roi from config
        is_loading_black_roi = np.average(img[:700, 0:250]) < 4.0
        if is_loading_black_roi:
            return True
    return False

while 1:
    success = char.select_by_template("SHOP_PORTAL")
    success &= wait_for_loading_screen(screen, 2)
    if success:
        break
    else:
        mouse.move(800, 450, randomize=50, delay_factor=[0.7, 0.7])
```

Reply:
Now I've copied a bit of code and made shop reset more robust and tolerant to clicking the porta too fast.